sbi/src/tsm.rs

```diff
@@ -171,6 +171,85 @@ impl TvmCpuRegister {
     }
 }
 
+/// Identifies a register set in the vCPU shared-memory state area layout.
+#[repr(u16)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum RegisterSetId {
+    /// General purpose registers.
+    Gprs = 0,
+    /// HS-level CSRs related to trap handling.
+    HypervisorTrapCsrs = 2,
+}
+
+impl RegisterSetId {
+    /// Returns the `RegisterSetId` corresponding to `val`.
+    pub fn from_raw(val: u16) -> Result<Self> {
+        match val {
+            0 => Ok(RegisterSetId::Gprs),
+            1 => Ok(RegisterSetId::HypervisorTrapCsrs),
+            _ => Err(Error::InvalidParam),
+        }
+    }
+}
+
+/// Specifies the location of a particular register set in the vCPU shared-memory state area.
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct RegisterSetLocation {
+    /// The type of the register set, identified by the `RegisterSetId` enum.
+    pub id: u16,
+    /// The version of the register set structure.
+    pub version: u16,
+    /// The offset of the register set from the start of the vCPU's shared-memory state area.
+    pub offset: u32,
+}
+
+/// General purpose registers.
+///
+/// The TSM will write to these registers upon return from `TvmCpuRun` when:
+///  - The vCPU takes a store guest page fault in an emulated MMIO region.
+///  - The vCPU makes an ECALL that is to be forwarded to the host.
+///
+/// The TSM will read from these registers when:
+///  - The vCPU takes a load guest page fault in an emulated MMIO region.
+///  - The host calls `TvmFinalize`, latching the entry point argument (stored in 'A1') for the
+///    TVM's boot vCPU.
+#[repr(C)]
+#[derive(Default)]
+pub struct Gprs([u64; 32]);
+
+/// HS-level CSRs related to trap handling.
+#[repr(C)]
+#[derive(Default)]
+pub struct HypervisorTrapCsrs {
```

Comment:
So we are grouping CSRs based on path where it is being accessed. This may be bit confused as the gprs or VS registers will probably be based on type.

May be define it in terms of context ?

GPR - Guest CPU Context
HypervisorTrapCsrs - Trap context
HTIMEDELTA - TIME Context
VS mode CSRs & HVIP - load_put_sync context

This aligns well with KVM. The downside is that other hypervisor may save/restore particular CSRs at different context.

How about just grouping them based on type ?

GPR, HMODE, VSMODE, SMODE

Reply:
> How about just grouping them based on type ?
> 
> GPR, HMODE, VSMODE, SMODE

Changed to group by spec/extension (and added a comment). I went the other way at first because having a bunch of register sets with only 1 or 2 CSRs seemed like overkill. But I agree that this grouping is more extensible.